framework/LearningGate.py

```diff
@@ -173,6 +217,393 @@ def train(self,trainingSet):
+#      weights[f] = weight
+#    # normalize weights
+#    ## METHOD: sum of weights should be unity
+#    scale = np.sum(weights)
+#    ## METHOD: by volume, assuming all weights are 1.0 initially before preference
+#    # vol = np.product(list(np.max(v) for v in clusterFeatureDict.values()))
+#    # print('DEBUGG original volume:',vol)
+#    # renormalize the entirety of the space to have the same hypervolume as before weighting
+#    # newVolume = np.product(weights)
+#    # oldVolume = 1.0 # because we scaled between 0 and 1, this will fail if you don't
+#    # scale = (oldVolume/newVolume)**(1.0/float(len(features)))
+#    ## END by volume
+#    for feature,vals in clusterFeatureDict.items():
+#      clusterFeatureDict[feature] = vals * scale
+#      v = clusterFeatureDict[feature]
+#      print('DEBUGG val range: {:15.15s} {:1.3e} {:1.3e} {:1.3e}'.format(feature,np.min(v),np.average(v),np.max(v)))
+#    vol = np.product(list(np.max(v) for v in clusterFeatureDict.values()))
+#    print('DEBUGG volume:',vol)
+#    return clusterFeatureDict
+
+  def _trainSubdomainRoms(self, templateRom, counter, trainingSet, strategy):
+    """
+      Trains the ROMs on each clusterable subdomain, and calculates features based on the data, rom
+      @ In, templateRom, SupervisedLearning instance, base ROM as a template for training
+      @ In, counter, list(tuple), instructions for subdividing domain into subdomains
+      @ In, trainingSet, dict, data on which ROMs should be trained
+      @ In, strategy, str, clustering strategy (e.g. "segment", "continuous", "cluster")
+      @ Out, clusterFeatureDict, dict, clustering information as {feature: [rom values]}
+      @ Out, roms, np.array(SubpervisedLearning instances), trained ROMs for each subdomain
+    """
+    targets = templateRom.target[:]
+    targets.remove(templateRom.pivotParameterID)
+    clusterFeatureDict = {}
+    roms = []
+    if self.targetDatas is None:
+      self.targetDatas = [] # DEBUGG only
+    for i,subdiv in enumerate(counter):
+      # slicer for data selection
+      picker = slice(subdiv[0], subdiv[-1]+1)
+      ## TODO only consider one sample at 0? Should do more for non-ARMA ROMs!
+      ##  -> for now, only ARMAs can be used with this method, so we can address this later
+      data = dict((var,[trainingSet[var][0][picker]]) for var in trainingSet)
+      targetData = dict((var,data[var][0]) for var in targets)
+      self.targetDatas.append(targetData) # DEBUGG only
+      # create a new ROM and train it
+      newRom = copy.deepcopy(templateRom)
+      self.raiseADebug('Training segment',i,picker)
+      newRom.train(data)
+      roms.append(newRom)
+      # if clustering, evaluate metrics
+      if strategy in ['continuous','cluster']:
+        ## -> basic metrics (mean, variance, etc)
+        basicData = self._evaluateBasicMetrics(targetData)
+        for feature, val in basicData.items():
+          if i == 0:
+            clusterFeatureDict[feature] = [val]
```

Comment:
I could; I would like to preserve the ability to check for the existence of features that haven't been included without automatically filling a value, but I'll keep this in mind in case it becomes more obviously beneficial.